pkg/network/encoding/http.go

```diff
@@ -168,49 +102,62 @@ func (e *httpEncoder) buildAggregations(payload *network.Connections) {
 			}
 
 			staticTags |= s.StaticTags
-
-			// It is a map to aggregate the same tag
-			if len(s.DynamicTags) > 0 {
-				if dynamicTags == nil {
-					dynamicTags = make(map[string]struct{})
-				}
-
-				for _, dynamicTag := range s.DynamicTags {
-					dynamicTags[dynamicTag] = struct{}{}
-				}
+			for _, dynamicTag := range s.DynamicTags {
+				dynamicTags[dynamicTag] = struct{}{}
 			}
 		}
 
-		e.staticTags[key.ConnectionKey] = staticTags
-		e.dynamicTagsSet[key.ConnectionKey] = dynamicTags
+		e.aggregations.EndpointAggregations = append(e.aggregations.EndpointAggregations, ms)
+	}
+
+	serializedData, _ := proto.Marshal(e.aggregations)
+	return serializedData, staticTags, dynamicTags
+}
 
-		aggregation.EndpointAggregations = append(aggregation.EndpointAggregations, ms)
+func (e *httpEncoder) Close() {
+	if e == nil {
+		return
 	}
+
+	e.reset()
+	e.byConnection.Close()
 }
 
 func (e *httpEncoder) getDataMap(stats map[uint16]*http.RequestStat) map[int32]*model.HTTPStats_Data {
-	res := make(map[int32]*model.HTTPStats_Data, len(stats))
+	resPtr := httpStatusCodeMap.Get().(*map[int32]*model.HTTPStats_Data)
+	e.toRelease = append(e.toRelease, resPtr)
+
+	res := *resPtr
 	for key := range stats {
 		res[int32(key)] = httpStatsDataPool.Get().(*model.HTTPStats_Data)
 	}
 	return res
 }
 
-func (e *httpEncoder) Close() {
+func (e *httpEncoder) reset() {
 	if e == nil {
 		return
 	}
-	for _, elem := range e.aggregations {
-		if elem == nil {
-			continue
-		}
-		for _, entry := range elem.EndpointAggregations {
-			for _, value := range entry.StatsByStatusCode {
-				httpStatsDataPool.Put(value)
-			}
-			httpStatsPool.Put(entry)
+
+	byEndpoint := e.aggregations.EndpointAggregations
+	for i, endpointAggregation := range byEndpoint {
+		byStatus := endpointAggregation.StatsByStatusCode
+		for k, s := range byStatus {
+			s.Reset()
+			httpStatsDataPool.Put(s)
+			delete(byStatus, k)
```

Comment:
Would just reallocating the map after the for loop be better (cpu-wise)?

Reply:
I haven't tried that to be honest since I was looking mostly at memory

Reply:
I could perhaps move the key deletion out of for loop and have this instead:

```
for k, s := range byStatus {
	s.Reset()
	httpStatsDataPool.Put(s)
	delete(byStatus, k)
}

for k := range byStatus {
	delete(byStatus, k)
}
```

Since the second loop is an idiom recognized by the Go compiler and doesn't really result in a regular iteration
https://github.com/golang/go/issues/20138

Reply:
That sounds fine :+1:

Reply:
Done